lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs

```diff
@@ -0,0 +1,150 @@
+        [JsonConstructor]
+        internal MetricsData(string serializedMetricsString)
+        {
+            var metrics = JsonConvert.DeserializeObject<IList<MetricTracker>>(serializedMetricsString, settings);
+
+            foreach (var m in metrics)
+            {
+                _metricsMap.TryAdd(m.GetMetric().Name, m);
+            }
+        }
+
+        public void RegisterMetric(IMetric metric)
+        {
+            if (!_metricsMap.TryAdd(metric.Name, new MetricTracker(metric)))
+            {
+                throw new ArgumentException("The metric [{0}] already exists.", metric.Name);
+            }
+        }
+
+        public bool TryGetMetric(string name, out IMetric me)
+        {
+            if (!_metricsMap.TryGetValue(name, out MetricTracker tracker))
+            {
+                me = null;
+                return false;
+            }
+            me = tracker.GetMetric();
+            return true;
+        }
+
+        public IEnumerable<MetricTracker> GetMetricTrackers()
+        {
+            return _metricsMap.Values;
+        }
+
+        /// <summary>
+        /// Flushes changes since last sink for each metric. 
+        /// Called when Driver is sinking metrics.
+        /// </summary>
+        /// <returns>Key value pairs of metric name and record that was flushed.</returns>
+        public IEnumerable<KeyValuePair<string, MetricTracker.MetricRecord>> FlushMetricRecords()
+        {
+            // for each metric, flush the records and create key value pairs
+            return _metricsMap.SelectMany(kv => kv.Value.FlushChangesSinceLastSink().Select(r => new KeyValuePair<string, MetricTracker.MetricRecord>(kv.Key, r)));
+        }
+
+        /// <summary>
+        /// Updates metrics given another <see cref="MetricsData"/> object.
+        /// For every metric in the new set, if it is registered then update the value,
+        /// if it is not then add it to the registration.
+        /// </summary>
+        /// <param name="metrics">New metric values to be updated.</param>
+        internal void Update(IMetrics metrics)
+        {
+            foreach (var tracker in metrics.GetMetricTrackers())
+            {
+                _metricsMap.AddOrUpdate(tracker.GetMetric().Name, tracker, (k, v) => v.UpdateMetric(tracker));
+            }
+        }
+
+        /// <summary>
+        /// Flushes that trackers contained in the queue.
+        /// Called when Evaluator is sending metrics information to Driver.
+        /// </summary>
+        /// <returns>Queue of trackers containing metric records.</returns>
+        internal IEnumerable<MetricTracker> FlushMetricTrackers()
+        {
+            return new ConcurrentQueue<MetricTracker>(_metricsMap.Select(kv => new MetricTracker(kv.Value.GetMetric(), kv.Value.ChangesSinceLastSink, (ConcurrentQueue<MetricTracker.MetricRecord>)kv.Value.FlushChangesSinceLastSink(), kv.Value.KeepUpdateHistory)));
```

Comment:
long line! also, we can get rid of that cast if we make `MetricTracker` constructor accept `IEnumerable` as parameter